### ISSUE TITLE
Comparison of Deepspeed Stage 1,2 and 3 vs DDP

### ISSUE THREAD

**Issue by jpatel-bdai** [microsoft/DeepSpeed#4815]
**Describe the bug**
When the model fits on a single GPU, how does Deepspeed ZeRO stage 1 compare with DDP? In my experiments, the Deepspeed ZeRO stage 1. I see that my overall loss training progresses similarly in both the cases but after a few iterations, the Deepspeed ZeRO stage 1 and stage 2 performance degrades. 

**Expected behavior**
I would expect both DDP and Deepspeed ZeRO Stage 1 to give similar results when run of single GPU. The total loss is a combination of a few losses and one of which is trans loss. Do you have experiments that compare DDP and Deepspeed ZeRO stage 1 or 2 that I can refer. Are these supposed to give similar performance? The attached screenshots are for single GPU and 2 GPU experiments for total loss and trans loss. 

**Screenshots**
![image](https://github.com/microsoft/DeepSpeed/assets/135647158/3d9c7cce-fc27-4fdb-b627-206190971f56)
![image](https://github.com/microsoft/DeepSpeed/assets/135647158/91b9d015-ca0e-4ec0-9087-ac09e8f6bf96)
![image](https://github.com/microsoft/DeepSpeed/assets/135647158/ae28b6f5-7325-4ce7-b0df-890cc41c21e2)
![image](https://github.com/microsoft/DeepSpeed/assets/135647158/bbdd99a7-a37f-454b-85d7-7f24a8ed6b8c)

**System info (please complete the following information):**
 - OS: [e.g. Ubuntu 20.04]
 - A100s  -> singe GPU  and 2 - GPU
 - Python version - 3.10
 
**Docker context**
No

**Comment by tjruwase** [comment]
@jpatel-bdai, all zero stages are expected to match ddp on single gpu runs. So, it appears that you are hitting bugs in zero. 

Are you able to share detailed steps to help us repro? Thanks!

**Comment by jpatel-bdai** [comment]
I will try to share the detailed steps to reproduce if possible. I am using the pytorch-lightning's [Deepspeed Strategy](https://lightning.ai/docs/pytorch/stable/api/lightning.pytorch.strategies.DeepSpeedStrategy.html). However, are all zero stages are expected to match ddp on multi-gpu runs as well?

**Comment by tjruwase** [comment]
Ideally, we expect zero stages to match ddp in multi-gpu runs, since zero is designed to be a memory-efficient ddp algorithm. In terms of debugging, a first step would be to inspect the training loss of each forward pass to detect deviations.

**Comment by chiragjn** [comment]
Don't want to hijack this issue, but I noticed that my train loss values are wildly different between stage 2 and stage 3, is that expected? I take that minor differences can happen because of different optimizer implementations but the differences in my case is too severe - I checked that everything was seeded the same way and with multiple restarts of stage 2 and stage 3 results were not exact but consistent with the same stage but not across

**Comment by olegsinavski** [comment]
Hello, I'm debugging the same issue. Since I'm working on VLMs, I found that the inclusion of the vision part (e.g. a `timm` model) leads to slower loss, but if I train regular LLMs, both converge well.
Here is a VLM training:
<img width="714" alt="image" src="https://github.com/microsoft/DeepSpeed/assets/2086260/d544016a-43a6-462b-a34d-c5776d908cff">
Here is an LLM training:
<img width="800" alt="image" src="https://github.com/microsoft/DeepSpeed/assets/2086260/f4636724-cea1-4af9-a96b-d5984fa9d5fc">

I discarded many options already: e.g. I fixed BatchNorms, used bf16-true, added "amp" decorators, used the same optimizers, disabled schedulers but I still see differences in the "vision" case. There are few other differences I'm trying to work out...

**Comment by tohtana** [comment]
Hi @jpatel-bdai,
We recently fixed multiple accuracy issues (e.g. #5104, #5105, #5150, and #5170). Some users reported that mismatches of loss values were solved by updating from 0.12.* to 0.14.*.
Can you try the latest version if you haven't?

**Comment by tohtana** [comment]
@jpatel-bdai Let me share my [verification script](https://github.com/tohtana/validate_zero). 

As long as I set FP32, PyTorch's Adam, and NP=2, it showed exact matches with PyTorch.
Currently this does not work well with FP16/BF16. I would appreciate it If you have any idea for improvement.

**Comment by tohtana** [comment]
Let me close this issue as we haven't had a new report for a while. Please feel free to reopen it if you still see the issue.

**Comment by EmaadKhwaja** [comment]
@tohtana Do you still see it not working well with FP16/BF16?

**Comment by EmaadKhwaja** [comment]
@tohtana Think this issue with worth reopening
